# Hypothesis Testing

The purpose of the test is to tell if there is any significant difference between two data sets.



## Overview

This module covers,

1) One sample and Two sample t-tests

2) ANOVA

3) Type I and Type II errors

4) Chi-Squared Tests

## Question 1 

*A student is trying to decide between two GPUs. He want to use the GPU for his research to run Deep learning algorithms, so the only thing he is concerned with is speed.*

*He picks a Deep Learning algorithm on a large data set and runs it on both GPUs 15 times, timing each run in hours. Results are given in the below lists GPU1 and GPU2.*

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
from scipy import stats 
import numpy as np

In [0]:
GPU1 = np.array([11,9,10,11,10,12,9,11,12,9,11,12,9,10,9])
GPU2 = np.array([11,13,10,13,12,9,11,12,12,11,12,12,10,11,13])

#Assumption: Both the datasets (GPU1 & GPU 2) are random, independent, parametric & normally distributed

Hint: You can import ttest function from scipy to perform t tests 

**First T test**

*One sample t-test*

Check if the mean of the GPU1 is equal to zero.
- Null Hypothesis is that mean is equal to zero.
- Alternate hypothesis is that it is not equal to zero.

In [0]:
import  scipy.stats  as stats 

In [10]:
stats.ttest_1samp(GPU1,0)

Ttest_1sampResult(statistic=34.056241516158195, pvalue=7.228892044970457e-15)

**Answer** - Mean is not equal to zero.
Since absolute T stat is more than 2 and p value is less than 0.05
Reject null hypothesis and accept alternate hypothesis.

In [9]:
stats.ttest_1samp(GPU2,0)

Ttest_1sampResult(statistic=37.406459913272236, pvalue=1.969680571681882e-15)

## Question 2

Given,

Null Hypothesis : There is no significant difference between data sets

Alternate Hypothesis : There is a significant difference

*Do two-sample testing and check whether to reject Null Hypothesis or not.*

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

In [11]:
stats.ttest_ind(GPU2, GPU1)

Ttest_indResult(statistic=2.627629513471839, pvalue=0.013794282041452725)

**Answer** - As Pvalue(1.379%) < 5% we reject the Null hyphothesis. And accept the Alternate Hypothesis : There is a significant difference

## Question 3

He is trying a third GPU - GPU3.

In [0]:
GPU3 = np.array([9,10,9,11,10,13,12,9,12,12,13,12,13,10,11])

#Assumption: Both the datasets (GPU1 & GPU 3) are random, independent, parametric & normally distributed

*Do two-sample testing and check whether there is significant differene between speeds of two GPUs GPU1 and GPU3.*

#### Answer:

In [15]:
stats.ttest_ind(GPU1, GPU3)

Ttest_indResult(statistic=-1.4988943759093303, pvalue=0.14509210993138993)

**Answer** - As Pvalue (14.509%) > 5% . Null Hyphothesis cannot be rejected.We can say that the GPUs are effectively same speed

## ANOVA

## Question 4 

If you need to compare more than two data sets at a time, an ANOVA is your best bet. 

*The results from three experiments with overlapping 95% confidence intervals are given below, and we want to confirm that the results for all three experiments are not significantly different.*

But before conducting ANOVA, test equality of variances (using Levene's test) is satisfied or not. If not, then mention that we cannot depend on the result of ANOVA

In [0]:
import numpy as np

e1 = np.array([1.595440,1.419730,0.000000,0.000000])
e2 = np.array([1.433800,2.079700,0.892139,2.384740])
e3 = np.array([0.036930,0.938018,0.995956,1.006970])

#Assumption: All the 3 datasets (e1,e2 & e3) are random, independent, parametric & normally distributed

Perform levene test on the data

The Levene test tests the null hypothesis that all input samples are from populations with equal variances. Levene’s test is an alternative to Bartlett’s test bartlett in the case where there are significant deviations from normality.

source: scipy.org

#### Answer:

In [18]:
stats.levene(e1,e2,e3)

LeveneResult(statistic=2.6741725711150446, pvalue=0.12259792666001798)

**Answer** - As pvalue is greater than 5%, we cannot reject the null hyphothesis, so we can say the there is no significant variance in the datasets and can perform ANOVA test.

## Question 5

The one-way ANOVA tests the null hypothesis that two or more groups have the same population mean. The test is applied to samples from two or more groups, possibly with differing sizes.

use stats.f_oneway() module to perform one-way ANOVA test

In [19]:
stats.f_oneway(e1,e2,e3)

F_onewayResult(statistic=2.51357622845924, pvalue=0.13574644501798466)

**Answer** - As the Pvalue (13.57%) is greater the 5%, we cannot reject Null hyphothesis. Also we can conclude the means of the datasets are not significantly different

## Question 6

*In one or two sentences explain about **TypeI** and **TypeII** errors.*

#### Answer:

Type I of errors is also known as a "false positive" or "false hit". It describes a situation where you reject the null hypothesis when it is actually true.

Type II error is also known as a "false negative" or "miss". It describes a situation where you fail to reject the null hypothesis when it is actually false. 

## Question 7 

You are a manager of a chinese restaurant. You want to determine whether the waiting time to place an order has changed in the past month from its previous population mean value of 4.5 minutes. 
State the null and alternative hypothesis.

#### Answer:


If the population mean does not change from 4.5 mins it will be Null Hypothesis
so 

H0 : mu = 4.5

Alternative hyphotesis will be that the mu is not equal to 4.5

HA : mu not equal to 4.5

## Chi square test

## Question 8

Let's create a small dataset for dice rolls of four players

In [22]:
import numpy as np

d1 = [5, 8, 3, 8]
d2 = [9, 6, 8, 5]
d3 = [8, 12, 7, 2]
d4 = [4, 16, 7, 3]
d5 = [3, 9, 6, 5]
d6 = [7, 2, 5, 7]

dice = np.array([d1, d2, d3, d4, d5, d6])

dice

array([[ 5,  8,  3,  8],
       [ 9,  6,  8,  5],
       [ 8, 12,  7,  2],
       [ 4, 16,  7,  3],
       [ 3,  9,  6,  5],
       [ 7,  2,  5,  7]])

run the test using SciPy Stats library

Depending on the test, we are generally looking for a threshold at either 0.05 or 0.01. Our test is significant (i.e. we reject the null hypothesis) if we get a p-value below our threshold.

For our purposes, we’ll use 0.01 as the threshold.

use stats.chi2_contingency() module 

This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table

Print the following:

- chi2 stat
- p-value
- degree of freedom
- contingency



In [29]:
chi2_stat,p_val,dof,con =  stats.chi2_contingency(dice)
print("Chi2 Stat is - ", chi2_stat)
print("P-Value is - ", p_val)
print("Degree of Freedom is - ", dof)
print("Contingency is - " , con)

Chi2 Stat is -  23.315671914716496
P-Value is -  0.07766367301496693
Degree of Freedom is -  15
Contingency is -  [[ 5.57419355  8.20645161  5.57419355  4.64516129]
 [ 6.50322581  9.57419355  6.50322581  5.41935484]
 [ 6.73548387  9.91612903  6.73548387  5.61290323]
 [ 6.96774194 10.25806452  6.96774194  5.80645161]
 [ 5.34193548  7.86451613  5.34193548  4.4516129 ]
 [ 4.87741935  7.18064516  4.87741935  4.06451613]]


**Answer** - As we can see in the above results the pvaule 7.766% is greater than the threshold of 1%. Null hyphothesis cant be rejected.

## Question 9

### Z-test

Get zscore on the above dice data using stats.zscore module from scipy. Convert zscore values to p-value and take mean of the array.

In [31]:
z_score = stats.zscore(dice)
p_val = stats.norm.sf(abs(z_score))
p_val.mean()

0.23428473233691496

## Question 10

A Paired sample t-test compares means from the same group at different times.

The basic two sample t-test is designed for testing differences between independent groups. 
In some cases, you might be interested in testing differences between samples of the same group at different points in time. 
We can conduct a paired t-test using the scipy function stats.ttest_rel(). 

In [0]:
before= stats.norm.rvs(scale=30, loc=100, size=500) ## Creates a normal distribution with a mean value of 100 and std of 30
after = before + stats.norm.rvs(scale=5, loc=-1.25, size=500)

In [33]:
print(before.shape)
print(after.shape)

(500,)
(500,)


Test whether a weight-loss drug works by checking the weights of the same group patients before and after treatment using above data.

In [34]:
stats.ttest_rel(a=before,b=after)

Ttest_relResult(statistic=4.470394607330399, pvalue=9.672341625637274e-06)

**Answer**  - As the pvalue is significantly less than 5% . We can say that weight loss drug works by checking the weights of the same group of patients before and after the treatment.